### ISSUE TITLE
[CN] Crash on array initialization

### ISSUE THREAD

**Issue by septract** [rems-project/cerberus#239]
CN version: e62d18a

Crashing input: 
```
void a() { int b[] = {0}; }
```

Result: 

```
% cn array_crash_1.c
other location (Cabs_to_ail.constructValue)  error: TODO: WellTyped infer_pexpr

PEctor <other location (Cabs_to_ail.constructValue)>
|-Array
`-Pexpr(TODO) <other location (Cabs_to_ail.constructValue)>
  `-Array(conv_loaded_int(''signed int'', a_475))

cn: internal error, uncaught exception:
    Failure("TODO: WellTyped infer_pexpr")
    Raised at Stdlib.failwith in file "stdlib.ml", line 29, characters 17-33
    Called from Dune__exe__WellTyped.BaseTyping.infer_pexpr in file "backend/cn/wellTyped.ml", line 1600, characters 22-29
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.pure in file "backend/cn/typing.ml", line 109, characters 22-25
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.pure in file "backend/cn/typing.ml", line 109, characters 22-25
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.pure in file "backend/cn/typing.ml", line 109, characters 22-25
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.pure in file "backend/cn/typing.ml", line 109, characters 22-25
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.pure in file "backend/cn/typing.ml", line 109, characters 22-25
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.bind in file "backend/cn/typing.ml", line 85, characters 8-11
    Called from Dune__exe__Typing.run in file "backend/cn/typing.ml", line 41, characters 16-19
    Called from Dune__exe__Main.main in file "backend/cn/main.ml", line 183, characters 20-87
    Called from Dune__exe__Main.main in file "backend/cn/main.ml", line 179, characters 9-1023
    Re-raised at Dune__exe__Main.main in file "backend/cn/main.ml", line 223, characters 8-73
    Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
    Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 34, characters 37-44
```

**Comment by scuellar** [comment]
The following spec also crashes:

```
extern int *createArray (unsigned int len);
/*@ spec createArray(pointer p, u32 len);
    requires true;
    ensures  take vp1 = each(u32 i; i < len) { Owned<int>( array_shift<int>(return, i)) };
@*/
```

I suspect these are related and also the same error behind #240 .

**Comment by cp526** [comment]
The above error is now fixed, so I'm closing this. 

Santiago's example is still broken, but this turns out to be unrelated – I opened another issue: in your example the `spec` for `createArray` and the function prototype mismatch in the argument list. Apparently whoever added support for CN's `spec` notation didn't put the required checks.

**Comment by cp526** [comment]
Fixed by 1bb6cd66e37bb5b73becbe7423cf3c2d749c12ef
